### PR TITLE
Nordic SoC: Add Nordic GPREGRET setting support

### PR DESCRIPTION
**Pull request by nordicjm**

### CODE REVIEW

File: soc/arm/nordic_nrf/nrf52/power.c [hunk 1]
```diff
@@ -115,3 +115,10 @@ bool _sys_soc_is_valid_power_state(enum power_states state)
 
 	return false;
 }
+
+/* Overrides the weak ARM implementation: set the general purpose retention register and reboot the module */
```
Comment (pdunaj) [inline soc/arm/nordic_nrf/nrf52/power.c]: Comment could be broken into two lines

**Review by pdunaj — APPROVED** [review]

**Review by ioannisg — CHANGES_REQUESTED** [review]
LGTM. Left a comment for using nrfx function.

File: soc/arm/nordic_nrf/nrf52/power.c [hunk 2]
```diff
@@ -115,3 +115,11 @@ bool _sys_soc_is_valid_power_state(enum power_states state)
 
 	return false;
 }
+
+/* Overrides the weak ARM implementation:
+   Set general purpose retention register and reboot */
+void sys_arch_reboot(int type)
+{
+	nrf_power_gpregret_set(type);
```
Comment (ioannisg) [inline soc/arm/nordic_nrf/nrf52/power.c]: I believe nrf_power_gpregret_set() takes an unsigned u8 as argument. So, please, check if passing -int- gives warning/error. Ideally, we should have type-casting here, right?

Comment (ioannisg) [reply soc/arm/nordic_nrf/nrf52/power.c]: Same comment applies for nRF51

**Comment by ioannisg** [comment]
The commit title should begin with "soc: nordic: "
And you need to add at least one "body" line in the commit message.

File: soc/arm/nordic_nrf/nrf51/power.c [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+ * Copyright (c) 2018 Intel Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <zephyr.h>
+#include <nrf_power.h>
+
+#define LOG_LEVEL CONFIG_SOC_LOG_LEVEL
+#include <logging/log.h>
+LOG_MODULE_DECLARE(soc);
```
Comment (carlescufi) [inline soc/arm/nordic_nrf/nrf51/power.c]: Do we need logging here? this file doesn't seem to be using it?

Comment (ioannisg) [reply soc/arm/nordic_nrf/nrf51/power.c]: Not really.

**Review by ioannisg — APPROVED** [review]

**Review by nashif — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

**Pull request merged**